rule BamPEFragmentSize:
    """
    This tool calculates the fragment sizes for read pairs given a BAM file from paired-end sequencing.
    Several regions are sampled depending on the size of the genome and number of processors to estimate
    the summary statistics on the fragment lengths. Properly paired reads are preferred for computation, i.e.,
    it will only use discordant pairs if no concordant alignments overlap with a given region.
    Is part of deeptools http://deeptools.readthedocs.io/en/latest/index.html

    Required input:
        __BamPEFragmentSize__input_bams: list of bam files

    Required output:
        __BamPEFragmentSize__output: png file
    """

    input:
        bam = __BamPEFragmentSize__input_bams
    params:
        bs = __BamPEFragmentSize__binsize
    log:
        __BamPEFragmentSize__logs
    output:
        __BamPEFragmentSize__output
    threads: 4
    run:
        labels = " ".join(str(os.path.basename(label).split("_")[0]) for label in input.bam)
        shell("bamPEFragmentSize --bamfiles  {input.bam} --histogram {output}  --binSize {params.bs} -p {threads} --samplesLabel " + labels + "  2> {log}")
